qiskit_experiments/curve_analysis/guess.py

```diff
@@ -24,39 +24,71 @@
 from qiskit_experiments.exceptions import AnalysisError
 
 
-def frequency(x: np.ndarray, y: np.ndarray) -> float:
-    """Get frequency of oscillating signal.
+def frequency(
+    x: np.ndarray,
+    y: np.ndarray,
+    filter_window: int = 5,
+    filter_dim: int = 2,
+) -> float:
+    r"""Get frequency of oscillating signal.
+
+    First this tries FFT. If the true value is likely below the Nyquist frequency,
+    the function tries low frequency fit with
+
+    .. math::
+
+        f_{\rm est} = \frac{1}{2\pi {\rm max}\left| y \right|}
+            {\rm max} \left| \frac{dy}{dx} \right|
+
+    given :math:`y = A \cos (2\pi f x + phi)`. In this mode, y data points are
+    smoothed by a Savitzky-Golay filter to protect against outlier points.
 
     .. note::
 
         This function returns always positive frequency.
+        DC offset value should be removed before applying this function.
 
     Args:
         x: Array of x values.
         y: Array of y values.
+        filter_window: Window size of Savitzky-Golay filter. This should be odd number.
+        filter_dim: Dimension of Savitzky-Golay filter.
 
     Returns:
         Frequency estimation of oscillation signal.
     """
-    sampling_rates = np.unique(np.diff(x))
-    sampling_rates = sampling_rates[sampling_rates > 0]
+    sampling_interval = np.unique(np.diff(x))
+    sampling_interval = sampling_interval[sampling_interval > 0]
```

Comment:
This is handling for edge case. Since user can scan experimental parameter with arbitrary value list, it can be negative spacing. But seems like sorted by x is right approach here.

Reply:
Added comments https://github.com/Qiskit/qiskit-experiments/pull/385/commits/45579785c06e8ac07aaabc6d02e91db25ff3f725

Reply:
For example, if we run
```python
expr = CrossResonanceHamiltonian((0, 1), np.linspace(0, 700, 15))
expr.set_analysis_options(return_data_points=True)

data1 = expr.run(backend)
exp_data.block_for_results()
data2 = expr.run(backend, experiment_data=data1)
data2.analysis_results(-1).value["x|c=0"]["xdata"]
```
it returns
```
array([0.0e+00, 5.0e-08, 1.0e-07, 1.5e-07, 2.0e-07, 2.5e-07, 3.0e-07,
       3.5e-07, 4.0e-07, 4.5e-07, 5.0e-07, 5.5e-07, 6.0e-07, 6.5e-07,
       7.0e-07, 0.0e+00, 5.0e-08, 1.0e-07, 1.5e-07, 2.0e-07, 2.5e-07,
       3.0e-07, 3.5e-07, 4.0e-07, 4.5e-07, 5.0e-07, 5.5e-07, 6.0e-07,
       6.5e-07, 7.0e-07])
```
this is somewhat monotonic but two set of results are concatenated. So there is non-continuous point in between two results, i.e. `7.0e-07, 0.0e+00`. Perhaps we can consider several edge cases but interpolate function is smart enough to handle non ordered data set.

I also noticed this function often use interpolate option due to small numerical noise in diff function. Perhaps we need to rewrite this function.

Reply:
Added data format logic in https://github.com/Qiskit/qiskit-experiments/pull/385/commits/e386e75280eea20f2f301425cf5abe0df9d688ec We can move them to data processor in separate PR.

Reply:
e386e75 looks good. I don't think the interpolate function is smart enough. In the [documentation](https://numpy.org/doc/stable/reference/generated/numpy.interp.html), it says:

> The x-coordinate sequence is expected to be increasing, but this is not explicitly enforced. However, if the sequence xp is non-increasing, interpolation results are meaningless.